Automattic/jetpack
# Publicize: Add REST API, follow-up

Follow-up to #10398 addressing some of the comments in https://github.com/Automattic/jetpack/pull/10398#issuecomment-434172435

- [x] Move most of the endpoint code to something like _inc/lib/core-api/wpcom-endpoints/publiciize.php and sync this with WP.com
- [x] Switch to `wpcom/v2` namespace
- [ ] Rename some of the routes. For example, `/publicize/connections` does not just return a list of connections. It runs the connection tests for each one, which is expensive. Currently implemented in this PR: `GET /publicize/services`, `GET /publicize/connections`, `GET /publicize/connection-test-results`.
- [x] Move a post's Publicize Connection info in to the Post endpoint itself via `register_rest_field()`.
- [x] Expose post's Publicize Message info via REST API's meta API.

Everything is up for debate :)

#### Changes proposed in this Pull Request:

All API endpoint code is in `_inc/lib/core-api`.

* `GET /wpcom/v2/publicize/services`
* `GET /wpcom/v2/publicize/connections`
* `GET /wpcom/v2/publicize/connection-test-results`
* `GET /wp/v2/posts/:id` includes `jetpack_publicize_connections` and `meta.jetpack_publicize_message` data.

#### Testing instructions:

##### Automated tests
1. `phpunit --group=publicize`
2. `phpunit --group=rest-api`

##### Manual tests

1. On a connected Jetpack site, with Gutenberg installed.
2. Get to a Gutenberg page (new post)
3. Open the JS console. 
4. Run `wp.apiFetch({path:'/wpcom/v2/publicize/connections' }).then( console.log )` and expect to get a proper list of connections.
5. Run `wp.apiFetch({path:'/wpcom/v2/publicize/services' }).then( console.log )` and expect to get a proper list of compatible social services.
6. Run `wp.apiFetch({path:'/wpcom/v2/publicize/connection-test-results }).then( console.log )` and expect (a slow response containing) a list of connections and whether they are correctly working.
6. Save the post you're currently looking at to get an id. Let's say you get `77`.
5. Run `wp.apiFetch({path:'/wp/v2/posts/77?context=edit' }).then( console.log )` and expect to see the post's Publicize connection details in the `jetpack_publicize_connections` field and the Publicize message string in the `meta.jetpack_publicize_message` field.

#### Proposed changelog entry for your changes:

* None needed
